drivers/periph_common: dac: add dac_play()

benpicco opened the pull request:
### Contribution description

This adds an API function to play a buffer of Audio samples using a DAC.
Double buffered operation is supported by specifying a callback that will be called when the next buffer can be queued with `dac_play()`.

### Testing procedure

A example application is provided.
It will play a raw audio (8kHz, 8 Bit) snipped from the [1961 synthesized speech demo](https://www.youtube.com/watch?v=ebK4wX76RZ4) ~~on startup~~. (Can be disabled since the file is rather large with 17k)

It will also play sine, square and sawtooth waves with frequencies that can be set on the command line.

I tested this with the on-board speaker (connected to `P0.26`) of the `mcb2388` as well as with hooking up a headphone to the ANALOG (`PA02`) header of the `same54-xpro`.

### Issues/PRs references

~~requires #13902~~

maribu commented on tests/periph_dac_play/main.c:
```diff
@@ -0,0 +1,152 @@
+    mutex_unlock(arg);
+}
+
+static void _play_sin_sample(const dac_cfg_t *cfg, uint8_t b, uint8_t pitch)
+{
+    for (uint16_t i = 0; i < DAC_BUF_SIZE; ++i) {
+        buf[b][i] = (isin(i << pitch) + 4096) >> 5;
+    }
+
+    dac_play(buf[b], DAC_BUF_SIZE, cfg);
+    mutex_lock(cfg->cb_arg);
+}
```
Wouldn't a constant buffer in ROM containing one period of the sine be much easier to use? By specifying the sample rate you should be able to generate different notes.

benpicco replied:
Yea I also should clean this up.
My intention was to have an example for the callback function, for generating sound on the fly.
(Although usually the callback would rather be used to read from a file)

benpicco replied:
Only using a single buffer now, not racing the playback anymore.

maribu commented on drivers/periph_common/dac.c:
```diff
@@ -0,0 +1,108 @@
+void dac_play(const void *buf, size_t len, const dac_cfg_t *cfg)
+{
+    const unsigned timer_freq  = 1000000;
+    const unsigned sample_rate = cfg->sample_rate * 6; /* XXX why the fudge factor? */
```
Wouldn't it work more reliable to internally work with a constant sampling rate? You could convert internally by either skipping over data or or playing samples multiple times. E.g. add a `step` variable, and use `buf[(pos += step) >> 8]` to access the sample.

Putting a constant load on the system regardless of sample frequency would it make easy to create effects like playing the same sample with different pitches, as you could just specify a higher sampling frequency.

benpicco replied:
Sorry for the long delay, I didn't find the time to get back to this PR.

I think that's a good idea, that simplifies the API quite a bit, I guess I wanted to avoid introducing functions to the API, but it's clearer to split it into setup and play than to shove everything into one function.

benpicco replied:
I also figured out the fudge factor - turns out you have to set the 'project sample rate' in Audacity, otherwise re-sampling the tracks will be undone on export :sweat_smile: 
So the sample was still 44.1kHz, that's why it had to be played faster.

miri64 commented:
@benpicco ping?

maribu commented:
@benpicco: When this is ready for the next round of review, could you squash and ping me?

benpicco commented:
Squashed.
I removed the just-in-time audio generation and split the API as you suggested.

benpicco commented:
@maribu I marked the API as experimental.
Ideally the DAC is driven by DMA  and the API should allow for that.
However, if it turns out some change is needed to accommodate DMA, we can change it in the future.

maribu reviewed (commented):
It seems I forgot to hit the "submit" button after reviewing. The comment might now be out of date :-/

benpicco commented on drivers/include/periph/dac.h:
```diff
@@ -129,6 +154,40 @@ void dac_poweron(dac_t line);
+ * @param[in] buf           A buffer with (audio) samples
+ * @param[in] params        Playback parameters
+ */
+void dac_play(dac_t dac, const void *buf, size_t len);
```
Is direct digital synthesis really what we are doing here? 
Where is the numerically-controlled oscillator here?

benpicco replied:
How about `dac_buffer_play()` / `dac_buffer_stop()`?

maribu replied:
Calling any form of directly generating waveforms from digital values is actually pretty common: https://www.arduino.cc/reference/en/libraries/dds/, https://hackaday.com/2016/02/12/embed-with-elliot-audio-playback-with-direct-digital-synthesis/

But `dac_buffer_play()` and `dac_buffer_stop()` is also a good name.

benpicco replied:
Ok, I changed the name.
I also moved it to a module as that makes dependency resolution less awkward.

benpicco commented on drivers/include/periph/dac.h:
```diff
@@ -102,6 +103,30 @@ enum {
+ * @note    Will be called in interrupt context. Only use the callback to signal a
+ *          thread. Don't directly fill the sample buffer in the callback.
+ */
+typedef void (*dac_cb_t)(void *arg);
```
I also noticed an inconsistency, since we should *always* call the callback when the second buffer can be queue. That means also after the first call to `dac_play()` - otherwise we are not using the double buffering at all.
(Still have to test that latest change)

chrysn commented:
From my initial look over this:

* (How) Can the application detect that it ran into an underflow? (I'm not sure it needs to, yet, but from an "errors should never occur silently" point of view, it'd seem like a good idea.)
* AFAIR @bergzand had experiments around for USB sound cards (but I don't find any PR). Comparing what was done there might help see whether the proposed API is practical.
* DMA operation will probably be specific to the chips and need more than just a generic DMA peripheral -- but we can still go there in time. (As can hooking them up to external DACs, probably via I2S).
* If dac_dds_set_cb is allowed to be called while playing (as "on the fly" implies), it'll need atomics or a critical section around writing that. I'm not sure how relevant that function is in practice, though -- outside of demos, how often do you change the data source without at the same time allowing to change the data format (sample rate and bit depth), in which case you're right back at dac_dds_init.
    * If a full format switch is called for (as would be the case when enabling gapless playback, even though where most needed, gapless is between like-formatted data anyway), can the API (or can it be extended to -- I'd be happy with a rough roadmap) do that in a double-buffered fashion? (I think so, it'd just need to double-buffer bit format as well and call timer_set_periodic without RESET_ON_SET when switching buffers, but maybe it'd be more complicated.)

benpicco commented:
> - (How) Can the application detect that it ran into an underflow?

If `dac_dds_is_playing()` returns `false` when it's our turn to fill the buffer, we know an underrun occurred.

> - If dac_dds_set_cb is allowed to be called while playing (as "on the fly" implies), it'll need atomics or a critical section around writing that.

added

> - DMA operation will probably be specific to the chips and need more than just a generic DMA peripheral -- but we can still go there in time

Yea that's why I initially put this to `periph/` so we can have a per-cpu implementation overwriting the generic one.
However, there we can't depend on periodic timers 'only if no CPU implementation exists.
The whole mess with `rtt_rtc` showed that we currently have no mechanism for 'feature may be provided by CPU or as a software fallback', so I just moved it to a module instead.

> - If a full format switch is called for (as would be the case when enabling gapless playback, even though where most needed, gapless is between like-formatted data anyway), can the API (or can it be extended to -- I'd be happy with a rough roadmap) do that in a double-buffered fashion? 

That sure could be done, but I'd rather avoid making the implementation more complex 'just in case'.
I don't expect switching between formats to be a thing that will occur often.

chrysn commented:
> If dac_dds_is_playing() returns false when it's our turn to fill the buffer, we know an underrun occurred.

That'd still be hard to query atomically. How about `dac_dds_play` return the previous status of is_playing?

Agreed on the periph part and the "avoid making [it] more complex 'just in case'".

benpicco commented:
Good idea, I dropped the `dac_dds_is_playing()` and implemented your suggestion.

chrysn commented:
I think there was a small race condition left -- didn't find the button to make a comment with a patch that's applicable, so I tried pushing directly.

chrysn reviewed (commented):
I'm generally happy with this; two of the comments should be addressed (on DAC_LINE and the question around "is it forbidden to call". There's the broader question of how such devices are initialized better in general (around DAC_NUMOF), but as this is an experimental module and more about exploring the application-facing API, I don't expect this to be addressed in the initial PR.

Other than those, what I'm missing is a second-user test. Still looking into what it'd take to get this to run on one of my boards (typically: periph_timer_periodic and a PWM-as-DAC implementation), but if someone else can give it a run with the boards where it's OOTB, this can go to ACK rather soon.

benpicco commented on drivers/include/dac_dds.h:
```diff
@@ -0,0 +1,140 @@
+ *              is ready to be filled.
+ *              Do not do the actual buffer filling inside the callback as this
+ *              is called from the timer context that is used to play the samples.
+ *              Instead, use it to wake a thread that then provides the samples
```
Oh it seems like that comment is misleading.
`dac_dds_play()` should absolutely be called inside the callback if the buffer is ready. What I meant is that reading the data into the buffer (e.g. from a file on SD card or generated at run-time) should not be done inside the interrupt.

benpicco commented on drivers/dac_dds/dac_dds.c:
```diff
@@ -0,0 +1,159 @@
+
+    struct dac_ctx *ctx = arg;
+
+    const dac_t dac    = ctx - _ctx;
```
tbh it's also not immediately obvious why this pointer arithmetic works.
I just added the `dac` to the callback struct, it's only 4 additional bytes.

chrysn commented:
> I just added the `dac` to the callback struct, it's only 4 additional bytes.

At this point it may be worth considering to decouple them even further
and have the init function not take a dac_t ("what if it's not linear")
but just a a plain dac_dds index ("device"? pointer if user-allocated?
not sure), and have separate arguments in the DAC.

(This may also help in conceptualization once someone gets the crazy
idea to add a dac_dds_init_multichannel that'd take multiple DACs into a
single dac_dds to allow stereo, 5.1 or RGB sound).

benpicco commented:
Hmm that would also solve the problem of assigning a timer to the dac_dds

chrysn commented:
On the test side, I can report success (with 15529 and a still-a-bit-haphazard PWM-posing-as-DAC implementation) -- I got a slightly metallic but understandable "Hello, ladies and gentlemen".

The only patch in this PR's code I had to add was to introduce a split between the DAC numbering and the dac_dss array, which is necessary as the pwm_dac produces dac_t values like 0x800 (16bit "PWM 0", 8bit "the PWM has an 8bit range", 8 bit "write to channel 0"); practically that meant hard-coding the `_ctx[dac].dac = 0x800` line instead of `.dac = dac`.

My proposed resolution is to have a dac_dss_t that indexes into the internal array (for however many there might be configured, a good default is probably 1 independent of the number of DAC channels), to take both the dac_dss_t and the dac_t in at dac_dds_init, and to use dac_dds_t in all other functions. Then, the example would need to provide both the dac_dds_t value (which in the example can really be hard-coded to 0), and a dac_t that may need to be configured per platform (basically a "where is the speaker connected").

benpicco commented:
Isn't that what ae488635a33503abdc051291b66ac29a0370b8c5 already did?

chrysn commented:
Missed that, sorry; yeah, that looks just like it. That probably makes it a good point to squash, then I can rerun my tests on the final version without running the risk of missing yet another commit.

benpicco commented:
Ok, squashed :smiley:

chrysn reviewed (approved):
Previous review comments all addressed. It's well configurable now, and works on the thingy:52.

If you know of any widespread boards it works well with already (the mcb2388 appears to be discontinued, and was probably on the more pricey side back then), I'd appreciate if you could make this more visible somewhere (the ["projects/showcase" thread in meta](https://forum.riot-os.org/t/user-projects-showcase-category/121/11) comes to mind).

ACK, and thanks for implementing this!
